explauto/interest_model/riac.py

```diff
@@ -0,0 +1,940 @@
+                    if split_ratio > np.random.random():
+                        return self.less.sample(sampling_mode=['random'])
+                    else:
+                        return self.greater.sample(sampling_mode=['random'])
+                
+        elif sampling_mode[0] == 'greedy':
+            # Choose the leaf with the max progress
+            if self.leafnode:
+                return self.sample_bounds()
+            else:
+                lp = self.less.progress
+                gp = self.greater.progress
+                if gp > lp:
+                    return self.greater.sample(sampling_mode=['greedy'])
+                else:
+                    return self.less.sample(sampling_mode=['greedy'])
+                
+        elif sampling_mode[0] == 'epsilon_greedy':
+            # Choose the leaf with the max progress with probability (1-eps) and a random leaf with probability (eps).
+            eps = sampling_mode[1]
+            if eps > np.random.random():
+                return self.sample(sampling_mode=['random'])
+            else:
+                return self.sample(sampling_mode=['greedy'])
+                
+        elif sampling_mode[0] == 'softmax':
+            # Sample leaves with probabilities progress*volume and a softmax exploration (with a temperature parameter)
+            if self.leafnode:
+                return self.sample_bounds()
+            else:
+                temperature = sampling_mode[1]
+                
+                leaves = self.get_leaves()
+                #print leaves[0].leafnode
+                #print map(lambda leaf:leaf.progress, leaves)
+                progresses = np.array(map(lambda leaf:leaf.progress*leaf.volume(), leaves)) #by volume
+                progress_max = max(progresses)
+                probas = np.exp(progresses / (progress_max*temperature))
+                probas = probas / np.sum(probas)
+                
+                #print "Softmax", progresses, probas
+                leaf = leaves[np.where(np.random.multinomial(1, probas)==1)[0][0]]
+                return leaf.sample_bounds()
+            
+        else:
+            raise NotImplementedError(sampling_mode)
+            
+            
+    def progress_all(self):
+        """
+        Competence progress of the overall tree
+        
+        """
+        return self.progress_idxs(range(np.shape(self.get_data_x())[0] - self.progress_win_size, np.shape(self.get_data_x())[0]))
+    
+            
+    def progress_idxs(self, idxs):
+        """
+        Competence progress on points of given indexes
+        
+        """
+        if self.progress_measure == 'abs_deriv':
+            if len(idxs) <= 1:
+                return 0
+            else:
+                idxs = sorted(idxs)[- self.progress_win_size:]
+                #print "progress_idxs", idxs, self.get_data_c()[idxs]
+                return abs(np.cov(zip(range(len(idxs)), self.get_data_c()[idxs]), rowvar=0)[0, 1])
```

Comment:
Progress measure : absolute value of competence derivative (code from discrete_progress)